added error messages factory

andriikorotkov opened the pull request:
Improve database sources and destinations error messages
When users are setting up a database connector and fail (due to misconfiguration) they are often provided a non-descriptive error message

The idea of this PR is to map common SqlState code to some human readable messages like: "Incorrect password or user name", "Incorrect host or port"
Thus user may have more information what might be wrong with provided configurations

Examples how it will look on UI:
 - mysql fail database
![2022-08-22_18-27](https://user-images.githubusercontent.com/88329385/185961792-4cb62e82-fe04-4365-8c7e-9fac6cb2ba8d.png)

-postgres fail username
![2022-08-22_18-29](https://user-images.githubusercontent.com/88329385/185961902-026053e1-aa12-4619-b02e-32e3bad51c6b.png)

- snowflake fail role
![2022-08-22_18-32](https://user-images.githubusercontent.com/88329385/185962032-e312784d-11f9-48be-b477-c1b619c9da0e.png)

- oracle fail host
![2022-08-22_18-34](https://user-images.githubusercontent.com/88329385/185962153-79fa633a-7dad-47bb-9322-6a5b8a78be16.png)

- GCS fail bucket name
![2022-08-22_18-38](https://user-images.githubusercontent.com/88329385/185962258-70cd861b-04b1-427c-9fd6-cf8ec6ba2ea2.png)

andriikorotkov commented:
/test connector=connectors/source-mysql-strict-encrypt

andriikorotkov commented:
/test connector=connectors/source-mssql-strict-encrypt

andriikorotkov commented:
/test connector=connectors/source-mssql

andriikorotkov commented:
/test connector=connectors/source-mysql

andriikorotkov commented:
/test connector=connectors/source-oracle-strict-encrypt

andriikorotkov commented:
/test connector=connectors/source-mongodb-strict-encrypt

andriikorotkov commented:
/test connector=connectors/source-mongodb-v2

andriikorotkov commented:
/test connector=connectors/source-snowflake

andriikorotkov commented:
/test connector=connectors/source-mongodb-v2

VitaliiMaltsev commented on airbyte-config/config-models/src/main/java/io/airbyte/config/exception/ConnectionFileServiceErrorException.java:
```diff
@@ -0,0 +1,20 @@
+
+package io.airbyte.config.exception;
+
+public class ConnectionFileServiceErrorException extends RuntimeException {
```
i did not find usages of this exception. Do we need this class?

andriikorotkov replied:
valid remark. removed it

VitaliiMaltsev commented on airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/DefaultJdbcDatabase.java:
```diff
@@ -73,10 +75,17 @@ public <T> Stream<T> unsafeResultSetQuery(final CheckedFunction<Connection, Resu
+      final DatabaseMetaData metaData = connection.getMetaData();
+      return metaData;
+    } catch (SQLException e) {
+      if (Objects.isNull(e.getCause())) {
```
please add some comments here in which cases or for which connectors we have _cause==null_

andriikorotkov replied:
updated it

VitaliiMaltsev commented on airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/ErrorMessageFactory.java:
```diff
@@ -0,0 +1,29 @@
+
+public class ErrorMessageFactory {
+
+  private final static Map<ConnectorName, ErrorMessage> MAP = Map.of(
```
Please rename MAP to CONNECTOR_ERROR_MESSAGE_MAP

VitaliiMaltsev commented on airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/messages/ErrorMessage.java:
```diff
@@ -0,0 +1,41 @@
+public abstract class ErrorMessage {
+
+  private static final String DEFAULT_ERROR_MESSAGE = "State code: %s. \n Error Message: %s.";
+  protected final Map<String, ConnectionErrorType> CONSTANTS = new HashMap<>();
```
the same here. Please rename CONSTANTS to ERRORCODES_TYPES

andriikorotkov replied:
renamed it

VitaliiMaltsev commented on airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/Source.java:
```diff
@@ -36,4 +37,14 @@ public interface Source extends Integration {
+   *         is not specified
+   */
+  default ConnectorName getConnectorName() {
+    return ConnectorName.DEFAULT;
```
i'm not sure that it's right way to return DEFAULT instead of true connectors name. I would prefer return  MYSQL or POSTGRES, etc

VitaliiMaltsev commented on airbyte-integrations/connectors/destination-gcs/src/main/java/io/airbyte/integrations/destination/gcs/GcsDestination.java:
```diff
@@ -52,10 +56,21 @@ public AirbyteConnectionStatus check(final JsonNode config) {
       return new AirbyteConnectionStatus().withStatus(Status.SUCCEEDED);
+    } catch (final AmazonS3Exception e) {
+      LOGGER.error("Exception attempting to access the Gcs bucket: {}", e.getMessage());
+      LOGGER.error("Please make sure you account has all of these roles: " + EXPECTED_ROLES);
```
I don't think this log is needed here. Lack of permissions is a special case as one of the possible reasons for AmazonS3Exception being thrown, but we need to provide the exact reason

andriikorotkov replied:
removed it

andriikorotkov commented:
/test connector=connectors/source-mongodb-strict-encrypt

andriikorotkov commented:
/test connector=connectors/source-mongodb-strict-encrypt

andriikorotkov commented:
@grishick @ryankfu I have seen all your comments, fixed them and updated the error message handling in this pull request. Now the error message that the user will see will consist of the state code, the error code, if any, and the message that comes when the exception is processed.

grishick reviewed (commented):
Why does MongoDB need a separate MongoDBErrorMessage class that maps some Mongo error codes to error types, but other DBs don't need that?

ryankfu commented on airbyte-integrations/connectors/source-snowflake/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/SnowflakeJdbcSourceAcceptanceTest.java:
```diff
@@ -96,9 +98,34 @@ public AbstractJdbcSource<JDBCType> getJdbcSource() {
+    assertTrue(actual.getMessage().contains("State code: 28000; Error code: 200028;"));
   }
 
   @Override
```
As a side comment, @cgardens has a [draft PR](https://github.com/airbytehq/airbyte/pull/15866/files) that covers some tests not covered here, namely invalid schema and database. Let's try to keep parity with tests since it appears that Postgres and other connectors already cover some of these tests

Also, can you link where the error codes were found? It help to be able to verify where `200028`, `200011`, etc are mentioned in their own documentation. For instance, `390100` is not found within their documentation for [Federated Authentication & SSO Error Codes](https://docs.snowflake.com/en/user-guide/errors-saml.html) so curious how you found these codes?

andriikorotkov replied:
@ryankfu as for the tests - I don't think our tests will contradict each other. In some cases, they may cover the same case on different datasets. Therefore, I do not think that it can interfere.

andriikorotkov replied:
@ryankfu as for the codes - now I don't process them, but only take the state code either from the **exception** or from **cause of this exception**. This code that I output is returned to me in a `SQLException`.

ryankfu replied:
It seems the messaging was lost, the comment was not intended to say that Charles has a PR that would potentially conflict with the changes here. The message was that here are the list of tests that should be covered in each of the different connectors. So namely invalid:
* host
* port
* username
* password
* database
* schema

The reason for asking for how the error/state codes were found is because some of the mongodb tests are asserting something like 
```
assertTrue(airbyteConnectionStatus.getMessage().contains("State code: -4"));
```
However from the standpoint of debugging, if someone wanted to look up what mongo db state code is how can it be determined that they're the right state codes? For instance looking up state codes for mongodb I could only find [these state code](https://www.mongodb.com/docs/v6.0/reference/exit-codes/) but none of them match `-4`. It begs the question, how would these state codes be useful if documentation from mongodb doesn't match the codes in the test and what they mean?

andriikorotkov replied:
@ryankfu yes, I agree with you that in some cases posting the code may not give results, but in most cases it is useful information that really helps the user. I can't tell you exactly why the driver developers do this (this is about mongodb), or why not all state codes are in the documentation (this is about your snowflake comment), but for most connectors all state codes are in the documentation and up a normal description is attached to them. This code is intended for general processing - and it handles messages from different connectors in the same way. Unfortunately, I cannot guarantee the accuracy of the state code, or the accuracy of the exception message.

grishick commented on airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/messages/MongoDbErrorMessage.java:
```diff
@@ -0,0 +1,29 @@
+
+import io.airbyte.integrations.base.errors.utils.ConnectorName;
+
+public class MongoDbErrorMessage extends ErrorMessage {
```
Please add a comment linking to MongoDB doc page that explains these error codes

andriikorotkov replied:
@grishick I reworked exception handling for mongodb. The error code and error message will now be passed directly from mongo exceptions. As for the `-3 and -4` codes - I didn't find them in the documentation, but I see that these numbers are assigned as an error code in the MongoException class in Mongo driver. I am attaching a screenshot where it occurs.
![2022-08-25_15-32](https://user-images.githubusercontent.com/88329385/186665596-628a9932-9c54-40f4-9c03-59f8d591c73a.png)

ryankfu replied:
This is helpful to know though. I'm curious to ask, how useful would exposing `state code` if it seems like the `state codes` are arbitrarily assigned with no real perceptible value to surface them?

andriikorotkov commented:
/test connector=connectors/source-mongodb-v2

andriikorotkov commented:
/test connector=connectors/source-mongodb-v2

andriikorotkov commented:
/test connector=connectors/source-mysql-strict-encrypt

andriikorotkov commented:
/test connector=connectors/source-mysql

andriikorotkov commented:
/test connector=connectors/source-mssql-strict-encrypt

andriikorotkov commented:
/test connector=connectors/source-postgres-strict-encrypt

andriikorotkov commented:
/test connector=connectors/source-postgres

andriikorotkov commented:
/test connector=connectors/source-oracle

andriikorotkov commented:
/test connector=connectors/source-oracle-strict-encrypt

andriikorotkov commented:
/test connector=connectors/source-snowflake

andriikorotkov commented:
/test connector=connectors/source-mssql

andriikorotkov commented:
/test connector=connectors/source-mongodb-strict-encrypt

andriikorotkov commented:
/test connector=connectors/source-oracle-strict-encrypt

andriikorotkov commented:
/test connector=connectors/source-oracle-strict-encrypt

andriikorotkov commented:
/test connector=connectors/source-oracle-strict-encrypt

andriikorotkov commented:
/test connector=connectors/source-oracle-strict-encrypt

andriikorotkov commented:
/test connector=connectors/source-oracle-strict-encrypt

andriikorotkov commented:
/test connector=connectors/destination-redshift

andriikorotkov commented:
/test connector=connectors/destination-mysql

andriikorotkov commented:
/test connector=connectors/destination-mysql-strict-encrypt

andriikorotkov commented:
/test connector=connectors/destination-postgres-strict-encrypt

andriikorotkov commented:
/test connector=connectors/destination-postgres

andriikorotkov commented:
/test connector=connectors/destination-mongodb-strict-encrypt

andriikorotkov commented:
/test connector=connectors/destination-mongodb

andriikorotkov commented:
/test connector=connectors/destination-mysql-strict-encrypt

andriikorotkov commented:
/test connector=connectors/destination-redshift

andriikorotkov commented:
/test connector=connectors/destination-postgres

andriikorotkov commented:
/test connector=connectors/destination-mysql-strict-encrypt

grishick commented:
Looks like a couple of destinations are failing tests. If that's the case, a workaround could be to exclude these failing connectors from this PR.

bleonard commented on airbyte-integrations/connectors/destination-gcs/src/main/java/io/airbyte/integrations/destination/gcs/GcsDestination.java:
```diff
@@ -52,10 +56,18 @@ public AirbyteConnectionStatus check(final JsonNode config) {
 
       return new AirbyteConnectionStatus().withStatus(Status.SUCCEEDED);
+    } catch (final AmazonS3Exception e) {
+      LOGGER.error("Exception attempting to access the Gcs bucket: {}", e.getMessage());
```
Maybe `AWS` not `Gcs`
Also, please look for other capitalization errors.

andriikorotkov replied:
Thank you for your comments. Changed it.

andriikorotkov commented:
/test connector=connectors/destination-redshift

andriikorotkov commented:
/test connector=connectors/destination-mysql-strict-encrypt

ryankfu commented on airbyte-db/db-lib/src/main/java/io/airbyte/db/exception/ConnectionErrorException.java:
```diff
@@ -0,0 +1,49 @@
+  }
+
+  public String getExceptionMessage() {
+    return exceptionMessage;
```
How come this return value is not `this.exceptionMessage;`? Also since these values are not set they will default to `null`, no? Is that intended behavior?

andriikorotkov replied:
This is normal method behavior. It just has to return me the message that was set to it. But to prevent NPE, I added the exceptionMessage value by initialization everywhere in the constructors

ryankfu commented on airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/AbstractSourceConnectorTest.java:
```diff
@@ -133,6 +133,10 @@ protected ConnectorSpecification runSpec() throws WorkerException {
+    return runCheck(getConfig());
+  }
+
+  protected StandardCheckConnectionOutput runCheck(JsonNode config) throws Exception {
```
How is `JsonNode config` used here?

andriikorotkov replied:
Sorry, but these are not my changes, but github for some reason shows that they are mine
![2022-08-26_11-52](https://user-images.githubusercontent.com/88329385/186866536-d73e78f7-e8c6-4aab-badc-c38ed3ad3b3d.png)
.

andriikorotkov replied:
Well noted. Now I no longer use this logic. Deleted it.

ryankfu commented on airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/errors/messages/ErrorMessage.java:
```diff
@@ -0,0 +1,29 @@
+  private static String configMessage(String stateCode, int errorCode, String message) {
+    var stateCodePart = Objects.isNull(stateCode) ? "" : "State code: " + stateCode + "; ";
+    var errorCodePart = errorCode == 0 ? "" : "Error code: " + errorCode + "; ";
+    return stateCodePart + errorCodePart + "Message: " + message;
```
Can we potentially use a `String.format()` here? How come the code is using type inference with `var` since it's clear the string will always be of type string , no?

andriikorotkov replied:
Changed the generation of `errorMessage` using `String.format()`. And using `var` is my habit. Replaced it with String

ryankfu commented on airbyte-integrations/connectors/destination-mysql-strict-encrypt/src/test-integration/java/io/airbyte/integrations/destination/mysql/MySqlTestDataComparator.java:
```diff
@@ -30,7 +33,35 @@ protected boolean compareBooleanValues(String firstBooleanValue, String secondBo
+  }
+
+  @Override
+  protected boolean compareDateTimeValues(String expectedValue, String actualValue) {
```
How is this work relevant to the error messages work?

andriikorotkov replied:
this logic is added here in the strict-encrypt connector from the main connector. Here is the number of the pull request and its name in which it was done - `[15245] Destination-mysql: fixed normalization tests after changes in python part`. The developer who added these changes only added them to the main connector, but forgot about strict-encrypt. Therefore, the tests will fail here, which interferes with testing and publishing.

andriikorotkov commented:
/test connector=connectors/destination-redshift

ryankfu commented on airbyte-integrations/connectors/destination-postgres/src/test/java/io/airbyte/integrations/destination/postgres/PostgresDestinationTest.java:
```diff
@@ -132,6 +140,72 @@ void testDefaultParamsWithSSL() {
+    ((ObjectNode) config).put(JdbcUtils.PASSWORD_KEY, PASSWORD);
+    ((ObjectNode) config).put(JdbcUtils.DATABASE_KEY, DATABASE);
+
+    var destination = new PostgresDestination();
```
Can we keep consistency with naming similar to `PostgresDestinationTest#sanityTest`? The test there uses 
```java
final Destination destination = new PostgresDestionation()
```

andriikorotkov replied:
updated it

andriikorotkov commented:
@grishick the tests for some connectors fail not because of my changes, but because for most of them the changes were mergrd without running the tests. For example, for redshift-destination, no changes were added to secrets in Google Secret Manager. For mysql-destination-strict-encrypt - no changes were added to the transformation of types (when added to mysql-destination they were not added here) and several more connectors with small errors.

ryankfu commented on airbyte-integrations/connectors/source-snowflake/src/main/java/io.airbyte.integrations.source.snowflake/SnowflakeSource.java:
```diff
@@ -113,4 +107,10 @@ private JsonNode buildUsernamePasswordConfig(final JsonNode config, final String
     return Jsons.jsonNode(configBuilder.build());
   }
 
+  @Override
```
How come this method was moved since it seems like nothing was changed? Correct me if I'm wrong

andriikorotkov replied:
Sorry, but these are not my changes. I don't know why GitHub displays them
![2022-08-26_15-02](https://user-images.githubusercontent.com/88329385/186899291-83c6e316-f118-469a-9035-a2f64dbe2c26.png)
.

ryankfu reviewed (changes requested):
There's a few things that still need to be reviewed, the things that seem important are mainly on:
* removing the use of `var` especially in areas where the code explicitly defines a type
* provide a link to each of the connectors state code since it wouldn't be super useful to report a state code for a connector only for ourselves to have difficult finding these codes
* remove the `MySqlTestDataComparator` work since that does not seem relevant to the error messages work

If there's a viewpoint that wasn't clear, let's work together to get this over the finish line. There's a lot a great work here and would love for this to benefit our users

andriikorotkov commented:
/test connector=connectors/destination-redshift

andriikorotkov commented:
> * remove the `MySqlTestDataComparator` work since that does not seem relevant to the error messages work

@ryankfu because of this part, i can't publish my changes to handle messages. This part breaks the tests and makes the normal publishing process impossible. (I try to adhere to the concept, and first run the tests there to publish without turning off the tests)

andriikorotkov commented on airbyte-db/db-lib/src/main/java/io/airbyte/db/mongodb/MongoDatabase.java:
```diff
@@ -47,6 +49,9 @@ public MongoDatabase(final String connectionString, final String databaseName) {
       database = mongoClient.getDatabase(databaseName);
+    } catch (final MongoConfigurationException e) {
+      LOGGER.error(e.getMessage());
+      throw new ConnectionErrorException(String.valueOf(e.getCode()), e);
```
Well noted. Added the message from this exception to the constructor

andriikorotkov commented on airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/DefaultJdbcDatabase.java:
```diff
@@ -76,6 +78,15 @@ public DatabaseMetaData getMetaData() throws SQLException {
+        throw new ConnectionErrorException(e.getSQLState(), e.getErrorCode(), e.getLocalizedMessage(), e);
+      } else {
+        var cause = (SQLException) e.getCause();
+        throw new ConnectionErrorException(e.getSQLState(), cause.getErrorCode(), cause.getLocalizedMessage(), cause);
```
because I call cause multiple times on the next line. I think yes, it looks better. Also I changed `var` to `SQLException` in this case.

andriikorotkov commented on airbyte-integrations/connectors/destination-gcs/src/main/java/io/airbyte/integrations/destination/gcs/GcsDestination.java:
```diff
@@ -52,10 +56,18 @@ public AirbyteConnectionStatus check(final JsonNode config) {
       return new AirbyteConnectionStatus().withStatus(Status.SUCCEEDED);
+    } catch (final AmazonS3Exception e) {
+      LOGGER.error("Exception attempting to access the AWS bucket: {}", e.getMessage());
+      var messages = getDefaultErrorMessage(e.getErrorCode(), e);
```
Yes you're right. I changed the `var` here to `String`, made the `getDefaultErrorMessage` method `private` and will use the `getErrorMessage` method everywhere.

andriikorotkov commented:
/test connector=connectors/destination-redshift

andriikorotkov commented on airbyte-integrations/connectors/destination-mongodb/src/main/java/io/airbyte/integrations/destination/mongodb/MongodbDestination.java:
```diff
@@ -71,18 +78,35 @@ public AirbyteConnectionStatus check(final JsonNode config) {
       }
       return new AirbyteConnectionStatus().withStatus(AirbyteConnectionStatus.Status.SUCCEEDED);
+    } catch (final ConnectionErrorException e) {
+      var messages = getErrorMessage(e.getStateCode(), e.getErrorCode(), e.getExceptionMessage(), e);
```
Yes, that's right, it's possible. But in processing, in the `ErrorMessage.getErrorMessage` method, I'll process it, and if the error code is empty, it won't change the readability of the processed error message.

ryankfu replied:
Wouldn't that mean that you can be passing in an `uninitialized` int into `getErrorMessage`? That would cause an error cause in `MongoDbSource` you create a `new ConnectionErrorException(String exceptionMessage)` but that does not include an `errorCode`

When I tested this quick snipped out in my IDE this was giving my an error compiling which mimics how `getErrorMessage` works, correct me if I'm wrong
```java
  @Test
  void testUninitialzedIntPassed() {
    final int number;
    System.out.println(runTest("hi", number, "world"));
  }

  public String runTest(final String first, final int num, final String second) {
    final String temp = num == 0 ? "" : " TEST";
    return first + temp + second;
  };
```

andriikorotkov commented:
/test connector=connectors/destination-redshift

andriikorotkov commented:
/test connector=connectors/destination-postgres

andriikorotkov commented:
@ryankfu I merged changes related to `MySqlTestDataComparator` in a separate pull request. Now they shouldn't get here

ryankfu commented:
Work was completed in this ticket: https://github.com/airbytehq/airbyte/pull/16202
